### PR TITLE
Merge master into 2.0 branch

### PR DESCRIPTION
**Pull request by Cortys**
As discussed in #525 I now merged master into the 2.0 branch, converting all new callback based methods to "promisified" ones with as little change to the original implementation as possible.

I merged via multiple smaller merge commits to keep them a bit more comprehensible. Since there were quite a lot of conflicts to resolve, it's probably best if someone else checks whether everything still works as intended.

Hope this helps getting the 2.0 development started.

### CODE REVIEW

**Comment by Cortys** [comment]
Tests seem to fail on 4.0 and 0.12 due to missing Promise global in the test environment. Will look into this...

File: src/loki-fs-structured-adapter.js [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+     * Generator for constructing lines for file streaming output of db container or collection.
+     *
+     * @param {object=} options - output format options for use externally to loki
+     * @param {int=} options.partition - can be used to only output an individual collection or db (-1)
+     *
+     * @returns {string|array} A custom, restructured aggregation of independent serializations.
+     * @memberof LokiFsStructuredAdapter
+     */
+    LokiFsStructuredAdapter.prototype.generateDestructured = function*(options) {
+      var idx, sidx;
```
Comment (techfort) [inline src/loki-fs-structured-adapter.js]: as we are in an ES6 context, can we tidy this file up a bit? use `const` for all vars that are not reassigned

File: src/loki-fs-structured-adapter.js [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        options.partition = -1;
+      }
+
+      // if partition is -1 we will return database container with no data
+      if (options.partition === -1) {
+        // instantiate lightweight clone and remove its collection data
+        dbcopy = this.dbref.copy();
+
+        for(idx=0; idx < dbcopy.collections.length; idx++) {
+          dbcopy.collections[idx].data = [];
```
Comment (techfort) [inline src/loki-fs-structured-adapter.js]: use `Object.keys()` here..

File: src/loki-fs-structured-adapter.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+        yield dbcopy.serialize({
+          serializationMethod: "normal"
+        });
+
+        return;
+      }
+
+      // 'partitioned' along with 'partition' of 0 or greater is a request for single collection serialization
+      if (options.partition >= 0) {
```
Comment (techfort) [inline src/loki-fs-structured-adapter.js]: this condition is redundant, `options.partition` can only be >= -1, if it's not -1 then this block gets executed, am I getting this wrong?

File: src/loki-fs-structured-adapter.js [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+        var doccount,
+          docidx;
+
+        // dbref collections have all data so work against that
+        doccount = this.dbref.collections[options.partition].data.length;
+
+        for(docidx=0; docidx<doccount; docidx++) {
```
Comment (techfort) [inline src/loki-fs-structured-adapter.js]: again, i believe there is a more FP way to loop over this. Only loop with `for` when there is a need to break the loop

File: src/loki-fs-structured-adapter.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+     * @returns {Promise} a Promise that resolves after the database was loaded
+     * @memberof LokiFsStructuredAdapter
+     */
+    LokiFsStructuredAdapter.prototype.loadDatabase = function(dbname)
+    {
+      var instream = fs.createReadStream(dbname);
+      var outstream = new stream();
+      var rl = readline.createInterface(instream, outstream);
+      var self=this;
+
```
Comment (techfort) [inline src/loki-fs-structured-adapter.js]: `var`s begone :)

File: src/loki-fs-structured-adapter.js [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+          // it should single JSON object (a one line file)
+          if (self.dbref === null && line !== "") {
+            self.dbref = JSON.parse(line);
+          }
+        });
+
+        // when that is done, examine its collection array to sequence loading each
+        rl.on('close', function() {
+          if(typeof self.dbref !== 'object')
+            reject(new Error('Database could not be read.'));
```
Comment (techfort) [inline src/loki-fs-structured-adapter.js]: brackets around ifs

**Comment by Viatorus** [comment]
@Cortys  Well done! I will look into this.

@techfort I think, Clemens wanted to first include the feature changes to v2.0.0 and promis-ified the very needed. There can be much more be done for ES6.

For FP (function programming) loops I would be careful because javascript does not really optimize such loops. Standard `for (let i = 0; i < array.length; i++)` are faster.

**Comment by Cortys** [comment]
@techfort Thanks!

Regarding ES6 I completely agree and migrating should probably one of the first things to do with 2.0. I stopped myself from already migrating in this PR to keep it a merge instead of a complete refactoring.

Since there are many ways to use ES6 and code style right now already is pretty inconsistent it might be a good idea to just enforce an ES6 code style via a strict ESLint ruleset (enforcing arrow functions, const decrarations, etc...).

**Comment by Viatorus** [comment]
Yes, I think this is a very good starting point for refactoring.

**Comment by techfort** [comment]
Linting would be a good idea but we'd have to limit it to the files that
are written in ES6, which is a bit ugly. As for fp loops yes, we adopted
for loops because of the speed,  but that is really only relevant for
synchronous queries. An fs adapter can sacrifice that level of speed for
more robust code I'd think,  but like for everything else in my life I
don't have extreme feelings about either approach.

On Thu 12 Jan 2017, 21:09 Viatorus <notifications@github.com> wrote:

Yes, I think this is a very good starting point for refactoring.

—
You are receiving this because you were mentioned.


Reply to this email directly, view it on GitHub
<https://github.com/techfort/LokiJS/pull/528#issuecomment-272284289>, or mute
the thread

**Comment by Cortys** [comment]
@techfort Which files do you have in mind that would not be written in ES6 after migration?

Regarding for-loops I have no strong feelings either.

**Comment by obeliskos** [comment]
Excellent, thanks @Cortys! Feel free to change anything you want.  Maybe our dev dependencies should be upgraded to latest versions and old node versions can probably go.   If this cleanup needs to occur over multiple commits I would be fine with that... better to have good starting dependencies that struggle with old dependencies.   PhantomJS (headless webkit?), karma, jasmine, jshint, (uglify?), etc we will probably want.  Maybe node 6 (or latest LTS 6.9.4?) minimum for travis ci?  

Thanks again... @techfort, once this is upgrade is finalized will we promote it to its own GitHub repo as well?

**Comment by techfort** [comment]
@Cortys @Viatorus @obeliskos first of all a massive thank you - i haven't had time to dedicate to Loki in a long while and you made a fantastic job of it. I approved the PR because i realised that it would be much better to lay out a roadmap of full migration to ES2015 instead of breaking people's balls to tidy up a small part of the lib written in ES2015. Good work everybody, and @obeliskos yes we should promote this to its own repo. Is there a native git way to turn a branch into a repo?

**Comment by VladimirTechMan** [comment]
@techfort @obeliskos (cc @Cortys @Viatorus) I guess it should be the same as (or similar to) this suggested workflow:
https://stackoverflow.com/questions/9527999/how-do-i-create-a-new-github-repo-from-a-branch-in-an-existing-repo/9529847#9529847

(Also, wanted to second @techfort: @Cortys @Viatorus @obeliskos well done, guys!)

**Comment by obeliskos** [comment]
I think your fingerprints are all over that code as well @VladimirTechMan...  

@techfort, please grant all existing contributors (or more) contributor access when you set up new repo,  Vlad's link looks useful whever you get time for that

**Pull request merged**